ui/src/app/edge/live/common/grid/modal/modal.ts

```diff
@@ -1,53 +1,146 @@
+    })
+    return [];
   }
 
-  protected override onCurrentData(currentData: CurrentData): void {
+  public static generateModalMeterPhases(component: EdgeConfig.Component, translate: TranslateService, userRole: Role): FormlyFieldLine[] {
+    let fields: FormlyFieldLine[] = [];
 
-    this.grid.mode = currentData.allComponents["_sum/GridMode"]
-    let gridActivePower = currentData.allComponents['_sum/GridActivePower']
-    this.grid.buyFromGrid = gridActivePower > 0 ? gridActivePower : 0;
-    this.grid.sellToGrid = gridActivePower < 0 ? (gridActivePower * -1) : 0;
+    for (let phase of ['L1', 'L2', 'L3']) {
+      fields.push(
+        {
+          type: 'line',
+          name: translate.instant('General.phase') + " " + phase + " ",
+          nameSuffix: Utils.ADD_NAME_SUFFIX_FOR_GRIDSELL_OR_GRIDBUY(translate),
+          indentation: TextIndentation.SIMPLE,
+          channel: component.id + '/ActivePower' + phase,
+          children: [
+            Role.isAtLeast(userRole, Role.INSTALLER) && {
+              type: 'line-item',
+              channel: component.id + '/Voltage' + phase,
+              converter: Utils.CONVERT_TO_VOLT,
+              indentation: TextIndentation.SIMPLE,
+            },
+            Role.isAtLeast(userRole, Role.INSTALLER) && {
+              type: 'line-item',
+              channel: component.id + '/Current' + phase,
+              converter: Utils.CONVERT_TO_CURRENT,
+              indentation: TextIndentation.SIMPLE,
+            },
+            {
+              type: 'line-item',
+              channel: component.id + '/ActivePower' + phase,
+              converter: Utils.CONVERT_TO_GRIDSELL_OR_GRIDBUY_POWER,
+              indentation: TextIndentation.SIMPLE
+            }
+          ]
+        })
+    };
+
+    return fields;
   }
 
-  protected setTranslatedName = (power: number | null) => {
-    if (power == null || power == 0) {
-      return "";
+  public static generateGridModal(edgeId: string, edgeConfig: EdgeConfig, userRole: Role, translate: TranslateService): FormlyFieldConfig[] {
+
+    const hasMultipleGridMeters = edgeConfig.getComponentsImplementingNature("io.openems.edge.meter.api.AsymmetricMeter")
+      .filter(comp => comp.isEnabled && edgeConfig.isTypeGrid(comp)).length > 1;
+
+    let lines: FormlyFieldLine[] = [
+      {
+        type: 'line',
+        name: translate.instant("General.offGrid"),
+        channel: '_sum/GridMode',
+        filter: Utils.isGridModeOffGrid()
+      },
+      {
+        type: 'line',
+        name: translate.instant("General.gridBuyAdvanced"),
+        channel: '_sum/GridActivePower',
+        converter: Utils.CONVERT_TO_GRIDBUY_POWER,
+      },
+      {
+        type: 'line',
+        name: translate.instant("General.gridSellAdvanced"),
+        channel: '_sum/GridActivePower',
+        converter: Utils.CONVERT_TO_GRIDSELL_POWER,
+      },
+    ]
+
+    for (let [key, component] of Object.entries(edgeConfig.components)) {
```

Comment:
`key` is not used, so a normal loop over `edgeConfig.components` should be sufficient.